drivers/mcp2515/mcp2515.c

```diff
@@ -0,0 +1,412 @@
+                                   enum mcp2515_mode mode)
+{
+    DEBUG("mcp2515_set_mode: mode=%x\n", mode);
+
+    if (mode == MODE_UNKNOWN) {
+        return -1;
+    }
+
+    enum mcp2515_mode cur_mode = mcp2515_get_mode(dev);
+
+    if (cur_mode == mode) {
+        return mode;
+    }
+    if (cur_mode == MODE_SLEEP) {
+        mcp2515_wake_up(dev);
+    }
+
+    mcp2515_spi_bitmod(dev, MCP2515_CANCTRL, MCP2515_CANCTRL_REQOP_MASK,
+                       (uint8_t)mode);
+    int cnt = 0;
+
+    while (cur_mode != mode && cnt++ < 2) {
+        cur_mode = mcp2515_get_mode(dev);
+    }
+    return cur_mode;
+}
+
+void mcp2515_wake_up(candev_mcp2515_t *dev)
+{
+    dev->wakeup_src = MCP2515_WKUP_SRC_INTERNAL;
+    mcp2515_spi_bitmod(dev, MCP2515_CANINTF, MCP2515_CANINTF_WAKIF,
+                       MCP2515_CANINTF_WAKIF);
+    xtimer_usleep(osc_startup(dev));
+
+    uint8_t flag = mcp2515_get_irq(dev);
+
+    if (flag & INT_WAKEUP) {
+        DEBUG("wakeup, irq raised\n");
+        dev->wakeup_src = 0;
+        mcp2515_clear_irq(dev, INT_WAKEUP);
+    }
+}
+
+static int mcp2515_enable_irq(candev_mcp2515_t *dev, uint8_t irq)
+{
+    return mcp2515_spi_bitmod(dev, MCP2515_CANINTE, irq, irq);
+}
+
+int mcp2515_configure_bittiming(candev_mcp2515_t *dev)
+{
+    struct can_bittiming *tim = &(dev->candev.bittiming);
+    int res = 0;
+    uint8_t c;
+    enum mcp2515_mode mode;
+
+    DEBUG("mcp2515_configure_bittiming: brp=%" PRIu32 ", prop_seg=%" PRIu32
+          ", phase_seg1=%" PRIu32 ", phase_seg2=%" PRIu32 "\n", tim->brp,
+          tim->prop_seg,
+          tim->phase_seg1, tim->phase_seg2);
+
+    mode = mcp2515_get_mode(dev);
+    if (mode != MODE_CONFIG) {
+        if (mcp2515_set_mode(dev, MODE_CONFIG) != MODE_CONFIG) {
+            return -1;
+        }
+    }
+
+    c = ((tim->brp - 1) & 0x3F) | ((tim->sjw - 1) << 6);
+    mcp2515_spi_write(dev, MCP2515_CNF1, &c, 1);
+
+
+    mcp2515_spi_bitmod(dev, MCP2515_CNF2, MCP2515_CNF2_PRSEG_MASK |
+                       MCP2515_CNF2_PHSEG_MASK | MCP2515_CNF2_BTLMODE,
+                       MCP2515_CNF2_BTLMODE | (tim->prop_seg - 1) |
+                       ((tim->phase_seg1 - 1) << 3));
+    mcp2515_spi_bitmod(dev, MCP2515_CNF3,
+                       MCP2515_CNF3_PHSEG_MASK | MCP2515_CNF3_WAKFIL,
+                       (tim->phase_seg2 - 1) | MCP2515_CNF3_WAKFIL);
+
+    if (mode != MODE_CONFIG) {
+        mcp2515_set_mode(dev, mode);
+    }
+
+    return res;
+}
+
+int mcp2515_init_irqs(candev_mcp2515_t *dev)
+{
+    return mcp2515_enable_irq(dev,
+                              MCP2515_CANINTE_RX0IE |
+                              MCP2515_CANINTE_RX1IE |
+                                //MCP2515_CANINTE_TX0IE |
+                                //MCP2515_CANINTE_TX1IE |
+                                //MCP2515_CANINTE_TX2IE |
+                              MCP2515_CANINTE_ERRIE |
+                                //MCP2515_CANINTE_MERRE |
```

Comment:
Why the comments?

Reply:
These are other interrupts that are available for the MCP2515, but are currently not used in the driver. 
If someone in the future would like to change the behavior of the driver, they can uncomment these interrupts and they will be ready to use.

Reply:
If he does he will have to change the code anyway no?

Reply:
I don't think so, no. Uncommenting these out will cause them to be handled in the same way as the other events: the _can_event_callback will be triggered when they occur. Currently the default action for each of the events in the event_callback is to simply DEBUG_print the event name, but adding other logic to the events should be fairly trivial, since it is done in usercode. Not in drivercode. 

You can see an example in `tests/candev/main.c`.

Reply:
I still don't like it too much, but I won't insist. Your call.

Reply:
I will remove them :)